### PR TITLE
Google/o auth

### LINKED ISSUE #148: Implement OAuth Authentication with Passport.js
### Describe the feature

Description
Integrate OAuth authentication into the application using Passport.js. This will allow users to log in using their existing accounts from social media platforms **Google** improving user experience and simplifying the sign-up process.

Acceptance Criteria:

Configure Passport.js OAuth Strategy:
Set up strategies for popular OAuth providers for **Sign in with Google**.
Ensure users can log in using their accounts from these providers.

User Account Management:
Create or link user accounts in the database upon successful authentication.
Store necessary user details (e.g., name, email, profile picture) after OAuth login.

OAuth Routes:
Implement the following routes:
Login with Provider (GET /auth/google / ): Redirect to the OAuth provider for user authentication.
Callback Route (GET /auth/google/callback / ): Handle the response from the provider, complete authentication, and redirect users to the application.

Technical Requirements:
Use passport-google-oauth20, or other necessary packages for OAuth.
Ensure that appropriate scopes are requested during OAuth login to gather necessary user information.
Handle error responses from OAuth providers gracefully.
Tasks:

 Install necessary packages for OAuth strategies (e.g., passport-google-oauth20).
 Configure Passport.js for OAuth strategies in passport-config.js.
 Implement routes for initiating OAuth login and handling callbacks.
 Create user accounts based on OAuth authentication results.
 Test the OAuth flow with different providers to ensure functionality.

### Add ScreenShots

image

### Record

- [X] I agree to follow this project's Code of Conduct
- [X] I'm a gssoc-24-extd contributor
- [X] I want to work on this issue

### PR DESCRIPTION
**Pull request by samar12-rad**
Implemented Google Oauth 2.0 in the backend and frontend 

## Summary by CodeRabbit
close #148 
## Release Notes

- **New Features**
  - Added Google OAuth authentication for seamless login.
  - Introduced buttons for signing in and signing up with Google in the Login and Signup components.
  - Enhanced the authentication flow with JWT for admin and customer routes.
  - New contributors added to the README for improved visibility.

- **Bug Fixes**
  - Improved error handling and logging for authentication processes.

- **Documentation**
  - Enhanced API self-documentation with updated responses.

- **Chores**
  - Updated dependencies to support new authentication strategies.
  - Refactored code for consistency and readability across components.

### CODE REVIEW

**Comment by samar12-rad** [comment]
for issue #197

**Review by RamakrushnaBiswal — COMMENTED** [review]
code rabit changes please

File: backend/config/oauth.config.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    {
+      clientID: process.env.GOOGLE_CLIENT_ID || config.GOOGLE_CLIENT_ID,
+      clientSecret:
+        process.env.GOOGLE_CLIENT_SECRET || config.GOOGLE_CLIENT_SECRET,
+      callbackURL: "/auth/google/callback",
+    },
+    async (accessToken, refreshToken, profile, done) => {
+      try {
+        // Extract the email from Google profile
+        const email = profile.emails[0].value;
```
[the remark this answers is not in the document]
Comment (samar12-rad) [reply backend/config/oauth.config.js:17]: It is not possible that google profile doesnt provide emails

File: backend/config/oauth.config.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+const GoogleStrategy = require("passport-google-oauth20").Strategy;
+const passport = require("passport");
+const Customer = require("../models/customer.model"); // Adjust the path as needed
+const config = require("./secret"); // Import your secrets (client ID, client secret)
```
[the remark this answers is not in the document]
Comment (samar12-rad) [reply backend/config/oauth.config.js:4]: secret file doesnt contain any secrets, it is just a file to keep collection of all imports of secrets from .env

**Comment by samar12-rad** [comment]
@RamakrushnaBiswal done

**Review by RamakrushnaBiswal — APPROVED** [review]

**Pull request merged**